gremlin-dotnet/src/Gremlin.Net/Driver/Connection.cs

```diff
@@ -172,28 +170,35 @@ private string SaslArgument()
             return Convert.ToBase64String(authBytes);
         }
 
-        private void BeginSendingMessages()
+        private void Enqueue(RequestMessage message)
         {
-            if (Interlocked.CompareExchange(ref _writeInProgress, 1, 0) != 0)
-                return;
-            SendMessagesFromQueueAsync().Forget();
+            _writeQueue.Enqueue(message);
+
+            if (Interlocked.Increment(ref _writeInProgress) == 1)
```

Comment:
(nitpick) I don't have a strong opinion on this, but an (integer) flag that can only have two values to represent the two states seems simpler to me than a counter that is only compared to `0` and `1`.

Reply:
Only that it's wrong. The [send-while-loop may exit](https://github.com/apache/tinkerpop/pull/1114/files/e9926556abad5587d3c70c5f0d480d6cdd254563#diff-d97ba309649d42be4056825bb0bfc82eL195), meanwhile the queuing observes the flag to be 1, not starting the loop. After that, [the flag is reset](https://github.com/apache/tinkerpop/pull/1114/files/e9926556abad5587d3c70c5f0d480d6cdd254563#diff-d97ba309649d42be4056825bb0bfc82eL196), but it's too late. The work item may never be processed.

Reply:
It doesn't help IMO to include CompareExchange into the while loop. When it's decided to exit the loop, another queueing attempt may already have observed the queue to be running, and work may be lost.

That being sad, I'm not particularly awesome at lock free programming. I think we could also just use ordinary locks in all those places and be good.

Reply:
Yeah, never mind. I though that `Decrement` could be replaced by the previous version with `CompareExchange`, but that would exit the `while` loop after a single iteration. Since you have to evaluate the state of `_writeInProgress` in every iteration, it really needs to be a counter instead of a flag.
It's really a bit confusing that `_writeInProgress` is now effectively a counter and a status flag at the same time, but I also don't know how we could solve that differently.

Reply:
(My answer actually only addressed your first comment here as I wrote it before seeing your second comment.)

> That being said, I'm not particularly awesome at lock free programming. I think we could also just use ordinary locks in all those places and be good.

Me neither, but how could we implement this with ordinary locks without blocking multiple task waiting to get the lock? The current implementation ensures that a single task can send all messages in the queue and do that in the background.
But it would in general be good if @jorgebay could also review this as he definitely has more experience than I do in this area.

Reply:
No, it's not a status flag ever, it's only a counter. The comparison `== 1` just asserts whether the increment (= work item queue) is the first one when idle, and thus has to start the loop. Same for the check on 0. We must ensure that we really drained the queue and decremented the counter to zero to be able to leave the loop.

What's non optimal is that we now have two sources of truth of the count of work items: The queue itself and the counter.

Reply:
>Me neither, but how could we implement this with ordinary locks without blocking multiple task waiting to get the lock?

That's not a big problem, you don't have to go lock free to be async. Also, you don't have to block inside of locks, and still you can maintain async code. Stephen Cleary has a wonderful library of async synchronization primitives that could help a lot here, but I'm not sure whether you'd want to take a dependency on a 3rd party lib.